### ISSUE TITLE
Issues clearing form while ensuring correct validation

### ISSUE THREAD

**Issue by dannief** [jquense/react-formal#56]
Hi,

First, thanks for the great set of libraries. 

Second, I am currently having an issue with clearing a form after a programmatic submit (i.e calling the `submit()` function). I have a form that is controlled and after submit, i set the `value` prop of the `Form` component to `customerSchema .default()`. At first this did nothing to clear the form values, but I finally realized that I needed to set default values of the fields in the schema. This now worked to clear the form but with some quirks and issues with validation that I will describe below. I have also attached a [sample](https://github.com/jquense/react-formal/files/252636/react-formal-test.zip) with the full code.

TL;DR
How to properly clear a form by, while still enabling validation to function as expected?

I have the following schema:

```
const customerSchema = yup
  .object().shape({
    name: yup.string().default('')
      .required('please enter a name'),

    dateOfBirth: yup.date()
      .max(new Date(), "Are you a time traveler?!"),

    // Required field validation occurs when field is nullable
    // But setting model to default DOES NOT clear the field
    // colorId: yup.number().nullable()
    //   .required('Please select a color')

    // Required field validation DOES NOT occur when field is set to default number
    // But setting model to default DOES clear the field
    colorId: yup.number().default(-1)
      .required('Please select a color')
  });

```

1. When I define my colorId schema like so
```
colorId: yup.number().default(-1)
      .required('Please select a color')
```
and my select field default option like so (i.e. setting the default value to be the same as schema default): 
`<option value={-1}>Select a color...</option>`
This clears the field as expected when I set the model to default 
```
this.setState({
    model: customerSchema.default(),
    errors: {}
});

```
but required field validation does not occur .ie. I didn't see the error I expected

2. When I define my colorId schema like so. (Based on the example in the docs)
```
colorId: yup.number().nullable()
    .required('Please select a color')
```
and my select field default option like so: 
`<option value={null}>Select a color...</option>`
The field does not clear the field as expected, when I set the model to default. But required validation does trigger in this case. (Note I also get a warning from React in the console, about setting value to null instead of '' or undefined)

3. After successfully submitting and clearing the form using code in No. 1, I get an issue where the last field I touched before hitting submit gets validated again and displays the validation error. I guess maybe the validation is triggered again since the field is blurred? But how to avoid this?

4. This is probably unrelated to the above issues. But using the sample code from the docs, when I add a date field like so `<Form.Field id="dateOfBirth" name='dateOfBirth'/>`, I get the following error in the console: `Date.js:72 Uncaught TypeError: Cannot read property 'value' of undefined`. If i set the `type` to 'text' error goes away.

Please see sample below:
[react-formal-test.zip](https://github.com/jquense/react-formal/files/252636/react-formal-test.zip)

**Comment by supermoos** [comment]
I would also like to know how to properly reset / clear / empty a form programatically after say something like a submit.

**Comment by jquense** [comment]
you can reset a form programmatically by changing the `value` prop to your "reset" value.

**Comment by supermoos** [comment]
But wouldn't that require the form to be a controlled form, anyway to reset when only using `<Form defaultValue={myValues} />`

**Comment by jquense** [comment]
not any non-hacky ways, just like controlled inputs, you can technically use a `key` prop that changes on reset to trick react into thinking the form is a new component.

apart from having some really specific and niche performance concern tho controlling the form is the correct way to do it.

**Comment by supermoos** [comment]
Arh, that key trick is neat, though hacky yes. It was really more to avoid some of the overhead boilerplate code that comes for each form field with a controlled component.

**Comment by jquense** [comment]
You don't need to touch the form.fields just control the `Form` component itself. it's as much boilerplate as control a single input. Slightly more than none, but not nearly as bad as wiring each input up (that is one major point of react-formal)

**Comment by supermoos** [comment]
Arh, nice! Hmm, any examples that showcase this way of controlled react-formal?

**Comment by jquense** [comment]
iTs just like a controlled input:

```
render() {
  let onChange = (value) => this.setState({ value })
  return (
   <Form 
      value={this.state.value} 
      onChange={onChange} 
      // ... 
    /> 
  );
}
```

**Comment by supermoos** [comment]
Thanks for that quick example, def is a lot easier than managing each input field! Question though, Clearing the form state like in your code triggers a validation error on the e-mail input field right away. It's being reset to `{email: ''}` cause setting email to undefined results in a warning. Is there a better way to define the yup validation object that would not invalidate the e-mail field when no value has been input?

```
        this.formModelSchema = yup.object({
            name: yup.object({
                first: yup.string().required('First name is required'),
                middle: yup.string(),
                last: yup.string().required('Last name is required')
            }),
            email: yup.string().notOneOf([this.props.userEmail], 'Don\'t use your email address').email().required()
        });
```

**Comment by jquense** [comment]
what is the warning? reseting doesn't trigger validation, but you might get a validation if when clicking reset, you blur an input or something. You can also control the errors show, like `value`, with `errors` and `onError`

**Comment by supermoos** [comment]
Warning is:
```
Warning: `value` prop on `input` should not be null. Consider using the empty string to clear the component or `undefined` for uncontrolled components.
    in input (created by FormControl)
    in FormControl (created by FormInput)
    in div (created by FormGroup)
    in FormGroup (created by FormInput)
    in FormInput (created by ChildBridge)
    in ChildBridge (created by MessageTrigger)
    in MessageTrigger (created by MessageListener)
    in MessageListener (created by ChildBridge)
    in ChildBridge (created by Binding)
    in Binding (created by Field)
    in form (created by Form)
    in MessageContainer (created by MessageListener)
    in MessageListener (created by Form)
    in BindingContext (created by Form)
    in Form (created by Uncontrolled(Form))
...
```

**Comment by supermoos** [comment]
The warning was because my custom Form.Field Component wasn't checking the value assigned to <Input>.

**Comment by Marcin-Kapusta** [comment]
I have controlled form and I want to reset to initial value. Problem is when I'm using Form.FieldArray inside Form and yup schema is assigned to the Form then when I add some Fields to FieldArray using arrayHelpers.push
and then rest fields (reset value has less fields in array) then validation process is getting corrupted with error like that for example:

Error: Yup.reach cannot resolve an array item at index: 2, in the path: fields[2].label. because there is no value at that index.

Is there any way to tell that now form will have again initial array length?

**Comment by jquense** [comment]
@Marcin-Kapusta can you put together a simple example with codesandbox?
